File: algorithms/666. Path Sum IV/solution.cc
class Solution {
public:
    int pathSum(vector<int>& nums) {
        int s[5][8] = {0};
        for ( int n : nums ) {
            int r = n / 100;
            int c = n / 10 % 10 - 1;
            int v = n % 10;
            s[r][c] = s[r-1][c/2] + v;
        }
        int sum = 0;
        for ( int i = 1; i < 5; ++i )
            for ( int j = 0; j < 8; ++j )
                if ( i == 4 || s[i][j] && ! s[i+1][j*2] && ! s[i+1][j*2+1] )
                    sum += s[i][j];
        return sum;
    }
};

// 195 / 195 test cases passed.
// Status: Accepted
// Runtime: 3 ms

File: algorithms/666. Path Sum IV/solution.py
class Solution(object):
    def pathSum(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        s = collections.defaultdict(int)
        l = collections.defaultdict(int)
        for n in nums[::-1]:
            r, c, v = n / 100, n / 10 % 10, n % 10
            l[r, c] = max(1, l[r + 1, c * 2] + l[r + 1, c * 2 - 1])
            s[r, c] = s[r + 1, c * 2] + s[r + 1, c * 2 - 1] + l[r, c] * v
        return s[1, 1]

# 195 / 195 test cases passed.
# Status: Accepted
# Runtime: 46 ms
